File: e2e/ruby/ws_chat_server.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# WebSocket chat server using Spikard WebSocket support

# Add packages/ruby to load path
$LOAD_PATH.unshift(File.expand_path('../../packages/ruby/lib', __dir__))

require 'spikard'

# WebSocket handler for chat messages
class ChatHandler < Spikard::WebSocketHandler
  def initialize
    super
    @username = nil
  end

  def handle_message(message)
    msg_type = message['type']
    puts "[ChatHandler] Received message type: #{msg_type}"
    puts "[ChatHandler] Message data: #{message.inspect}"

    case msg_type
    when 'chatMessage'
      # Echo back the chat message
      {
        'type' => 'chatMessage',
        'username' => message['username'] || 'anonymous',
        'message' => message['message'] || '',
        'timestamp' => message['timestamp']
      }

    when 'userJoined'
      # Store username and send acknowledgment
      @username = message['username']
      {
        'type' => 'userJoined',
        'username' => @username,
        'timestamp' => message['timestamp']
      }

    when 'userLeft'
      # Send departure message
      username = @username || message['username'] || 'anonymous'
      {
        'type' => 'userLeft',
        'username' => username,
        'timestamp' => message['timestamp']
      }

    else
      # Unknown message type - return nil to not send response
      nil
    end
  end

  def on_connect
    puts '[ChatHandler] Client connected'
  end

  def on_disconnect
    puts "[ChatHandler] Client disconnected (user: #{@username})"
  end
end

# Create Spikard app
app = Spikard::App.new

# Register WebSocket endpoint
app.websocket('/chat') do
  ChatHandler.new
end

# Run server
if __FILE__ == $PROGRAM_NAME
  puts 'Starting WebSocket chat server on ws://localhost:8000/chat'
  config = Spikard::ServerConfig.new(host: '0.0.0.0', port: 8000)
  app.run(config: config)
end

File: e2e/python/ws_chat_server.py
#!/usr/bin/env python3
"""WebSocket chat server using Spikard WebSocket support"""

import asyncio
from typing import Any

from spikard import Spikard
from spikard.config import ServerConfig
from spikard.websocket import WebSocketHandler


class ChatHandler(WebSocketHandler):
    """WebSocket handler for chat messages"""

    def __init__(self):
        self.username: str | None = None

    async def handle_message(self, message: dict[str, Any]) -> dict[str, Any] | None:
        """Handle incoming chat messages"""
        msg_type = message.get("type")
        print(f"[ChatHandler] Received message type: {msg_type}")
        print(f"[ChatHandler] Message data: {message}")

        if msg_type == "chatMessage":
            # Echo back the chat message
            return {
                "type": "chatMessage",
                "username": message.get("username", "anonymous"),
                "message": message.get("message", ""),
                "timestamp": message.get("timestamp"),
            }

        elif msg_type == "userJoined":
            # Store username and send acknowledgment
            self.username = message.get("username")
            return {
                "type": "userJoined",
                "username": self.username,
                "timestamp": message.get("timestamp"),
            }

        elif msg_type == "userLeft":
            # Send departure message
            username = self.username or message.get("username", "anonymous")
            return {
                "type": "userLeft",
                "username": username,
                "timestamp": message.get("timestamp"),
            }

        # Unknown message type - return None to not send response
        return None

    async def on_connect(self) -> None:
        """Called when a client connects"""
        print("[ChatHandler] Client connected")

    async def on_disconnect(self) -> None:
        """Called when a client disconnects"""
        print(f"[ChatHandler] Client disconnected (user: {self.username})")


# Create Spikard app
app = Spikard()


# Register WebSocket endpoint
@app.websocket("/chat")
def chat_endpoint():
    """WebSocket endpoint for chat"""
    return ChatHandler()


if __name__ == "__main__":
    print("Starting WebSocket chat server on ws://localhost:8000/chat")
    config = ServerConfig(host="0.0.0.0", port=8000)
    app.run(config=config)
